### PR TITLE
Unzip nested .zip file for manually uploaded zip-archives

### PR DESCRIPTION
**Pull request by anton-abushkevich**

### CODE REVIEW

File: src/main/java/com/odysseusinc/arachne/portal/service/submission/impl/BaseSubmissionServiceImpl.java [hunk 1]
```diff
@@ -549,27 +550,43 @@ public void uploadCompressedResultsByDataOwner(Long submissionId, File compresse
...
-            }
+            unzipRecursively(new ZipFile(compressedFile), unzipDir, MAX_NESTING_FOR_UNZIP_INNER_ZIP_FILES);
             storeExtractedFiles(submission, unzipDir, user.getId());
         } finally {
             FileUtils.deleteDirectory(unzipDir.toFile());
         }
     }
 
+    private void unzipRecursively(final ZipFile zipFile, final Path unzipDir, final int nesting) throws IOException {
+        final List<FileHeader> fileHeaders = zipFile.getFileHeaders().stream()
```
Comment (wivern) [inline src/main/java/com/odysseusinc/arachne/portal/service/submission/impl/BaseSubmissionServiceImpl.java:563]: If some zip files reside inside nested directory, they wouldn't be unzipped

File: src/main/java/com/odysseusinc/arachne/portal/service/submission/impl/BaseSubmissionServiceImpl.java [hunk 2]
```diff
@@ -549,27 +550,43 @@ public void uploadCompressedResultsByDataOwner(Long submissionId, File compresse
...
+        for (FileHeader fileHeader : fileHeaders) {
+            final String relativeFilePath = fileHeader.getFileName();
+            final String relativePath = FilenameUtils.getPath(relativeFilePath);
+            if(isNotBlank(relativePath)){
+                unzipDir.resolve(relativePath).toFile().mkdirs();
+            }
+            final File localFile = unzipDir.resolve(relativeFilePath).toFile();
+            FileUtils.copyInputStreamToFile(zipFile.getInputStream(fileHeader), localFile);
+
+            if (relativeFilePath.endsWith(".zip") && nesting > 0) {
```
Comment (wivern) [inline src/main/java/com/odysseusinc/arachne/portal/service/submission/impl/BaseSubmissionServiceImpl.java:575]: Why we should limit the level of archive nesting? 
Anyway, what is the reason to execute the procedure when the limit is exceeded?

**Review by wivern — CHANGES_REQUESTED** [review]
I'd like to suggest completely review the algorithm.
- there should not be limits, when using right algorithm there's no risk to get an infinite loop because in any case it's a finite file tree
- we can replace recursion with queue and/or visitor pattern, going this way there's no risk to reach stack overflow

File: src/main/java/com/odysseusinc/arachne/portal/service/submission/impl/BaseSubmissionServiceImpl.java (hunk 1 above)
Comment (anton-abushkevich) [reply src/main/java/com/odysseusinc/arachne/portal/service/submission/impl/BaseSubmissionServiceImpl.java:562]: they would

File: src/main/java/com/odysseusinc/arachne/portal/service/submission/impl/BaseSubmissionServiceImpl.java (hunk 2 above)
Comment (anton-abushkevich) [reply src/main/java/com/odysseusinc/arachne/portal/service/submission/impl/BaseSubmissionServiceImpl.java]: agreed. nesting limit removed

**Comment by anton-abushkevich** [comment]
> 
> 
> I'd like to suggest completely review the algorithm.
> 
>     * there should not be limits, when using right algorithm there's no risk to get an infinite loop because in any case it's a finite file tree
> 
>     * we can replace recursion with queue and/or visitor pattern, going this way there's no risk to reach stack overflow

Limit removed. Recursion is replaced with queue.

**Review by wivern — APPROVED** [review]

**Pull request merged**